### ISSUE TITLE
(evil-replace-state) from evil-mode does not replace characters

### ISSUE THREAD

**Issue by Hi-Angel** [Wilfred/remacs#831]
`evil-mode` is a vim-mode for emacs. That `(evil-replace-state)` is when in "normal mode" you click <kbd>Shift</kbd>-<kbd>r</kbd>, and successive strokes start replacing the text under caret.

It works in Emacs, but in REmacs successive strokes *insert text before caret*, instead of replacing it. Tested with latest `evil-mode` and clean config.

# Steps to reproduce:

1. In a directory without any `.emacs` configs *(they might get destroyed)* execute:
	```bash
	$ git clone https://github.com/emacs-evil/evil
	…
	$ echo "(add-to-list 'load-path \"$(pwd)/evil\")    (require 'evil)    (evil-mode 1)" > .emacs
	$ env HOME=$(pwd) remacs
	```
	
2. In the popped up `remacs` switch to `*scratch*` buffer, press <kbd>Shift</kbd>-<kbd>r</kbd>, and start typing

# Expected

Text under caret being replaced with the one that's being typed

# Actual

The text being typed piles up before the caret.

# Additional info

remacs build: 2018-07-27

I can take a stab at fixing this, but I need some pointers for what to look at, because I'm clueless how to debug it.

**Comment by shaleh** [comment]
I think you can put the .emacs bits into a test-evil-mode.el and load it after starting up with `remacs -Q`. This should help making it safely reproducible.

If you can work out a way to put the invocation of evil and the checking of the result into a .el file too then we have a test script. We could then use git bisect to find which commit broke this.

**Comment by Hi-Angel** [comment]
> We could then use git bisect to find which commit broke this.

Do you mean it's a regression, it worked before?

**Comment by shaleh** [comment]
If it works in emacs then remacs broke it when we ported something. So we can walk the commits backwards looking for the change.

Another possibility is this was fixed in emacs and we are missing the fix.

**Comment by shaleh** [comment]
When submitting issues here, it helps to be explicit about the emacs version tested. That way we know if it is before or after our fork and if we have brought in fixes since.

**Comment by db48x** [comment]
It's even simpler than this; the ordinary `overwrite-mode` is broken as well. When I hit the insert key to turn on `overwrite-mode` I get this message:

    funcall-interactively: One of nil, overwrite-mode-textual or overwrite-mode-binary should be specified: overwrite-mode-textual

This is odd, because it looks like it _is_ setting it to `overwrite-mode-textual`.

**Comment by Hi-Angel** [comment]
Ok, I made git-bisect, it shows:


	commit 8fe09389f592f888b51bb8e5f1ddbcdeb949ede9
	Author: Daniel Brooks <db48x@db48x.net>
	Date:   Wed May 23 11:33:37 2018 -0700

		port store_symval_forwarding

	:040000 040000 73653b7a43069c81743a929ed5553d44dee9a678 330cb1247152e56ab7345e13bc46c53f331a3f31 M      rust_src
	:040000 040000 77346b1e67aaf7041b818e03e362bcd49684569d 2d223362a4c9352775037496694ce39fd68f857b M      src

CC: @db48x

**Comment by shaleh** [comment]
Would you share the test you used?

**Comment by Hi-Angel** [comment]
> Would you share the test you used?

As was noted in previous comment, it's reproducible with plain "overwrite mode" that can be enabled with <kbd>Insert</kbd> — so that's what I was using for testing. That said, I checked later that `evil-replace-state` works before the commit too.

If you're asking for a script — I didn't make one, because offhand I didn't know how to make an external script to change a buffer, check something, and then mak <i>(r)</i>emacs to exit with specific code based on that, and I deemed that time-wise it was better to do bisection manually.
